crates/wac-types/src/targets.rs

```diff
@@ -0,0 +1,122 @@
+use crate::{ExternKind, ItemKind, SubtypeChecker, Types, WorldId};
+use std::collections::{BTreeMap, BTreeSet};
+use std::fmt;
+
+/// The result of validating a component against a target world.
+pub type TargetValidationResult = Result<(), TargetValidationReport>;
+
+/// The report of validating a component against a target world.
+#[derive(Debug, Default)]
+pub struct TargetValidationReport {
+    /// Imports present in the component but not in the target world.
+    pub imports_not_in_target: BTreeSet<String>,
+    /// Exports present in the target world but not in the component.
+    ///
+    /// This is a mapping from the name of the missing export to the expected item kind.
+    pub missing_exports: BTreeMap<String, ItemKind>,
+    /// Mismatched types between the component and the target world.
+    ///
+    /// This is a mapping from name of the mismatched item to a tuple of
+    /// the extern kind of the item and type error.
+    pub mismatched_types: BTreeMap<String, (ExternKind, anyhow::Error)>,
+}
+
+impl From<TargetValidationReport> for TargetValidationResult {
+    fn from(report: TargetValidationReport) -> TargetValidationResult {
+        if report.imports_not_in_target.is_empty()
+            && report.missing_exports.is_empty()
+            && report.mismatched_types.is_empty()
+        {
+            Ok(())
+        } else {
+            Err(report)
+        }
+    }
+}
+
+impl fmt::Display for TargetValidationReport {
+    fn fmt(&self, f: &mut fmt::Formatter<'_>) -> fmt::Result {
+        if !self.imports_not_in_target.is_empty() {
+            writeln!(
+                f,
+                "Imports present in the component but not in the target world:"
+            )?;
+            for import in &self.imports_not_in_target {
+                writeln!(f, "  - {}", import)?;
+            }
+        }
+        if !self.missing_exports.is_empty() {
+            writeln!(
+                f,
+                "Exports required by target world but not present in the component:"
+            )?;
+            for (name, item_kind) in &self.missing_exports {
+                writeln!(f, "  - {}: {:?}", name, item_kind)?;
+            }
+        }
+        if !self.mismatched_types.is_empty() {
+            writeln!(
+                f,
+                "Type mismatches between the target world and the component:"
+            )?;
+            for (name, (kind, error)) in &self.mismatched_types {
+                writeln!(f, "  - {}: {:?} ({})", name, kind, error)?;
+            }
+        }
+        Ok(())
+    }
+}
+
+impl std::error::Error for TargetValidationReport {}
+
+/// Validate whether the component conforms to the given world
+pub fn validate_target(
```

Comment:
It might not be immediately obvious where the arguments are coming from. Can we add a code example:

```rust
let mut types = Types::default();

let mut resolve = wit_parser::Resolve::new();
let pkg = resolve.push_dir(path_to_wit_dir)?.0;
let wit_bytes = wit_component::encode(&resolve, pkg)?;
let wit = Package::from_bytes("wit", None, wit_bytes, &mut types)?;

let component_bytes = std::fs::read(path_to_component)?;  
let component = Package::from_bytes("component", None, component_bytes, &mut types)?;
let wit_world = get_wit_world(&types, wit.ty())?;

validate_target(&types, wit_world, component.ty())?;
```